---
ms.date: 07/09/2020
ms.topic: reference
title: Ошибки и исключения в системе расширенного типа
description: Ошибки и исключения в системе расширенного типа
ms.openlocfilehash: 295c16ad9abb67b0c4967bf32125bfc7ee0a35da
ms.sourcegitcommit: ba7315a496986451cfc1296b659d73ea2373d3f0
ms.translationtype: MT
ms.contentlocale: ru-RU
ms.lasthandoff: 12/10/2020
ms.locfileid: "92652469"
---
# <a name="errors-and-exceptions-in-the-extended-type-system"></a><span data-ttu-id="c3f43-103">Ошибки и исключения в системе расширенного типа</span><span class="sxs-lookup"><span data-stu-id="c3f43-103">Errors and exceptions in the Extended Type System</span></span>

<span data-ttu-id="c3f43-104">Ошибки могут возникать в ETS во время инициализации данных типа и при доступе к члену объекта **PSObject** или при использовании одного из служебных классов, таких как **LanguagePrimitives**.</span><span class="sxs-lookup"><span data-stu-id="c3f43-104">Errors can occur in ETS during the initialization of type data and when accessing a member of an **PSObject** object or using one of the utility classes such as **LanguagePrimitives**.</span></span>

## <a name="runtime-errors"></a><span data-ttu-id="c3f43-105">Ошибки среды выполнения</span><span class="sxs-lookup"><span data-stu-id="c3f43-105">Runtime errors</span></span>

<span data-ttu-id="c3f43-106">За одним исключением при приведении все исключения, вызываемые из ETS во время выполнения, являются исключением **екстендедтипесистемексцептион** или исключением, производным от класса **екстендедтипесистемексцептион** .</span><span class="sxs-lookup"><span data-stu-id="c3f43-106">With one exception, when casting, all exceptions thrown from ETS during runtime are either an **ExtendedTypeSystemException** exception or an exception derived from the **ExtendedTypeSystemException** class.</span></span> <span data-ttu-id="c3f43-107">Это позволяет разработчикам скриптов перехватывать эти исключения с помощью `Trap` инструкции в сценарии.</span><span class="sxs-lookup"><span data-stu-id="c3f43-107">This allows script developers to trap these exceptions using the `Trap` statement in their script.</span></span>

## <a name="errors-getting-member-values"></a><span data-ttu-id="c3f43-108">Ошибки при получении значений элементов</span><span class="sxs-lookup"><span data-stu-id="c3f43-108">Errors getting member values</span></span>

<span data-ttu-id="c3f43-109">Все ошибки, возникающие при получении значения элемента ETS (свойство, метод или параметризованное свойство) вызывают исключение **жетвалуиксцептион** или **жетвалуеинвокатионексцептион** .</span><span class="sxs-lookup"><span data-stu-id="c3f43-109">All errors that occur when getting the value of an ETS member (property, method, or parameterized property) cause a **GetValueException** or **GetValueInvocationException** exception to be thrown.</span></span>
<span data-ttu-id="c3f43-110">Когда ETS распознает, что произошла ошибка, возникает исключение **жетвалуиксцептион** .</span><span class="sxs-lookup"><span data-stu-id="c3f43-110">When ETS recognizes that an error occurred a **GetValueException** exception is thrown.</span></span> <span data-ttu-id="c3f43-111">Когда базовый метод получения ссылочного члена распознает, что произошла ошибка, возникает исключение **жетвалуеинвокатионексцептион** , которое может включать или не содержать внутреннее исключение, вызвавшее ошибку вызова Get.</span><span class="sxs-lookup"><span data-stu-id="c3f43-111">When the underlying getter of a referenced member recognizes that an error occurred, a **GetValueInvocationException** exception is thrown that may or may not include the inner exception that caused the get invocation error.</span></span>

## <a name="errors-setting-member-values"></a><span data-ttu-id="c3f43-112">Ошибки при установке значений элементов</span><span class="sxs-lookup"><span data-stu-id="c3f43-112">Errors setting member values</span></span>

<span data-ttu-id="c3f43-113">Все ошибки, возникающие при установке значения свойства ETS, вызывают исключение **сетвалуиксцептион** или **сетвалуеинвокатионексцептион** .</span><span class="sxs-lookup"><span data-stu-id="c3f43-113">All errors that occur when setting the value of an ETS property cause a **SetValueException** or **SetValueInvocationException** exception to be thrown.</span></span> <span data-ttu-id="c3f43-114">Когда ETS распознает, что произошла ошибка, возникает исключение **сетвалуиксцептион** .</span><span class="sxs-lookup"><span data-stu-id="c3f43-114">When ETS recognizes that an error occurred a **SetValueException** exception is thrown.</span></span> <span data-ttu-id="c3f43-115">Если базовый метод задания упоминаемого свойства распознает, что произошла ошибка, возникает исключение **сетвалуеинвокатионексцептион** , которое может включать или не содержать внутреннее исключение, вызвавшее ошибку вызова Set.</span><span class="sxs-lookup"><span data-stu-id="c3f43-115">When the underlying setter of a referenced property recognizes that an error occurred, a **SetValueInvocationException** exception is thrown that may or may not include the inner exception that caused the set invocation error.</span></span>

## <a name="errors-invoking-a-method"></a><span data-ttu-id="c3f43-116">Ошибки при вызове метода</span><span class="sxs-lookup"><span data-stu-id="c3f43-116">Errors invoking a method</span></span>

<span data-ttu-id="c3f43-117">Все ошибки, возникающие при вызове метода ETS, вызывают исключение **месодексцептион** или **месодинвокатионексцептион** .</span><span class="sxs-lookup"><span data-stu-id="c3f43-117">All errors that occur when invoking an ETS method cause a **MethodException** or **MethodInvocationException** exception to be thrown.</span></span> <span data-ttu-id="c3f43-118">Когда ETS распознает, что произошла ошибка, возникает исключение **месодексцептион** .</span><span class="sxs-lookup"><span data-stu-id="c3f43-118">When ETS recognizes that an error occurred a **MethodException** exception is thrown.</span></span> <span data-ttu-id="c3f43-119">Если метод, на который указывает ссылка, распознает, что произошла ошибка, возникает исключение **месодинвокатионексцептион** , которое может включать или не содержать внутреннее исключение, вызвавшее ошибку вызова.</span><span class="sxs-lookup"><span data-stu-id="c3f43-119">When the referenced method recognizes that an error occurred, a **MethodInvocationException** exception is thrown that may or may not include the inner exception that caused the invocation error.</span></span>

## <a name="casting-errors"></a><span data-ttu-id="c3f43-120">Ошибки приведения</span><span class="sxs-lookup"><span data-stu-id="c3f43-120">Casting errors</span></span>

<span data-ttu-id="c3f43-121">Если предпринимается недопустимое приведение, создается исключение **псинвалидкастексцептион** .</span><span class="sxs-lookup"><span data-stu-id="c3f43-121">When an invalid cast is attempted, an **PSInvalidCastException** is thrown.</span></span> <span data-ttu-id="c3f43-122">Так как это исключение является производным от **System. InvalidCastException**, его невозможно напрямую перехватывать из скрипта.</span><span class="sxs-lookup"><span data-stu-id="c3f43-122">Because this exception derives from **System.InvalidCastException**, it is not able to be directly trapped from script.</span></span> <span data-ttu-id="c3f43-123">Имейте в виду, что сущностям, пытающимся выполнить приведение, пришлось бы обернуть **псинвалидкастексцептион** в **псрунтимиксцептион** для того, чтобы они были перехвачены скриптами.</span><span class="sxs-lookup"><span data-stu-id="c3f43-123">Be aware that the entity attempting the cast would need to wrap **PSInvalidCastException** in an **PSRuntimeException** for this to be trappable by scripts.</span></span> <span data-ttu-id="c3f43-124">Если предпринимается попытка установить значение **пспропертисет**, **псмемберсет**, **псмесодинфо** или члена **реадонлипсмемберинфоколлектион ' 1**, создается исключение **NotSupportedException** .</span><span class="sxs-lookup"><span data-stu-id="c3f43-124">If an attempt is made to set the value of an **PSPropertySet**, **PSMemberSet**, **PSMethodInfo**, or a member of the **ReadOnlyPSMemberInfoCollection\`1**, a **NotSupportedException** is thrown.</span></span>

## <a name="common-runtime-errors"></a><span data-ttu-id="c3f43-125">Распространенные ошибки среды выполнения</span><span class="sxs-lookup"><span data-stu-id="c3f43-125">Common runtime errors</span></span>

<span data-ttu-id="c3f43-126">Любые другие распространенные ошибки времени выполнения относятся к типу **екстендедтипесистемексцептион** Exception без дополнительных конкретных типов исключений.</span><span class="sxs-lookup"><span data-stu-id="c3f43-126">Any other common runtime errors that occur are of type **ExtendedTypeSystemException** exception with no additional specific exception types.</span></span>

## <a name="initialization-errors"></a><span data-ttu-id="c3f43-127">Ошибки инициализации</span><span class="sxs-lookup"><span data-stu-id="c3f43-127">Initialization errors</span></span>

<span data-ttu-id="c3f43-128">При инициализации могут возникать ошибки `types.ps1xml` .</span><span class="sxs-lookup"><span data-stu-id="c3f43-128">Errors may occur when initializing `types.ps1xml`.</span></span> <span data-ttu-id="c3f43-129">Как правило, эти ошибки отображаются при запуске среды выполнения PowerShell.</span><span class="sxs-lookup"><span data-stu-id="c3f43-129">Typically, these errors are displayed when the PowerShell runtime starts.</span></span> <span data-ttu-id="c3f43-130">Однако они также могут отображаться при загрузке модуля.</span><span class="sxs-lookup"><span data-stu-id="c3f43-130">However, they can also be displayed when a module is loaded.</span></span>
